File: frontend/src/app/features/incidents/domain/model/incident.ts
import { IncidentId } from '../value-objects/incident-id';

export const incidentSeverities = ['Low', 'Medium', 'High', 'Critical'] as const;

export type IncidentSeverity = (typeof incidentSeverities)[number];
export type IncidentSeverityFilter = IncidentSeverity | 'All';
export type IncidentPriority = 'P1' | 'P2' | 'P3' | 'P4';
export type IncidentStatus = 'Open' | 'Acknowledged' | 'In Progress' | 'Resolved';

export interface Incident {
  readonly id: IncidentId;
  readonly title: string;
  readonly description: string;
  readonly location: string;
  readonly assetId: string;
  readonly reportedAt: string;
  readonly severity: IncidentSeverity;
  readonly priority: IncidentPriority;
  readonly status: IncidentStatus;
  readonly operationalSignals: readonly string[];
}

export interface NewIncident {
  readonly title: string;
  readonly description: string;
  readonly location: string;
  readonly assetId: string;
  readonly severity: IncidentSeverity;
  readonly priority: IncidentPriority;
  readonly operationalSignals: readonly string[];
}

export interface IncidentQuery {
  readonly searchTerm: string;
  readonly severity: IncidentSeverityFilter;
}

File: backend/src/main/java/com/infraflow/platform/incidents/domain/Incident.java
package com.infraflow.platform.incidents.domain;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Objects;

public record Incident(
  IncidentId id,
  String title,
  String description,
  String location,
  AssetId assetId,
  OffsetDateTime reportedAt,
  IncidentSeverity severity,
  IncidentPriority priority,
  IncidentStatus status,
  List<String> operationalSignals,
  Long version
) {

  public Incident {
    Objects.requireNonNull(id, "Incident id is required.");
    Objects.requireNonNull(assetId, "Asset id is required.");
    Objects.requireNonNull(reportedAt, "Reported time is required.");
    Objects.requireNonNull(severity, "Severity is required.");
    Objects.requireNonNull(priority, "Priority is required.");
    Objects.requireNonNull(status, "Status is required.");
    title = requireText(title, "Title is required.");
    description = requireText(description, "Description is required.");
    location = requireText(location, "Location is required.");
    operationalSignals = List.copyOf(operationalSignals == null ? List.of() : operationalSignals);
  }

  public Incident acknowledge() {
    IncidentWorkflowPolicy.ensureCanAcknowledge(this);
    return withStatus(IncidentStatus.ACKNOWLEDGED);
  }

  public Incident startResponse() {
    IncidentWorkflowPolicy.ensureCanStartResponse(this);
    return withStatus(IncidentStatus.IN_PROGRESS);
  }

  public Incident resolve() {
    IncidentWorkflowPolicy.ensureCanResolve(this);
    return withStatus(IncidentStatus.RESOLVED);
  }

  public Incident withStatus(IncidentStatus nextStatus) {
    return new Incident(
      id,
      title,
      description,
      location,
      assetId,
      reportedAt,
      severity,
      priority,
      nextStatus,
      operationalSignals,
      version
    );
  }

  private static String requireText(String value, String message) {
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException(message);
    }

    return value.trim();
  }
}
